src/box/engine.c

```diff
@@ -235,14 +235,16 @@ engine_prepare_join(struct engine_join_ctx *ctx)
 }
 
 int
-engine_join(struct engine_join_ctx *ctx, struct xstream *stream)
+engine_join(struct engine_join_ctx *ctx, struct engine_checkpoint_cursor *cur,
+	    struct xstream *stream)
 {
 	ERROR_INJECT_YIELD(ERRINJ_ENGINE_JOIN_DELAY);
 
 	int i = 0;
 	struct engine *engine;
 	engine_foreach(engine) {
-		if (engine->vtab->join(engine, ctx->array[i], stream) != 0)
+		void *ctx_object = ctx->array != NULL ? ctx->array[i] : NULL;
```

Comment:
I don't understand this change: `ctx->array` must be allocated because `engine_prepare_join` must be called first.

Reply:
I see that the problem is that you don't call `engine_prepare_join`. That's confusing. Can we call `engine_prepare_join` and `engine_complete_join` for both checkpoint and read view join?

Reply:
> Can we call engine_prepare_join and engine_complete_join for both checkpoint and read view join?

Done